### DISCUSSION TITLE
How do I reuse or negate a guard condition?

### DISCUSSION THREAD

**Discussion by smac89** [statelyai/xstate#3017] [Q&A] [answered]
Say I have declared a guard somewhere in the machine:

```ts
{
    guards: {
        isPrivateRoute: ({ isPrivateRoute }) => isPrivateRoute,
        isAuthenticated: ({ isAuthenticated, isFirebaseAuthReady }) => isAuthenticated && isFirebaseAuthReady,
    },
}
```

And somewhere else in the state, I have a conditional transition with a `cond` that should return:
```
guards.isAuthenticated && !guards.isPrivateRoute
```
Is that possible to do? If not, what would be the best way to go about doing that?

**Comment by smac89** [comment, marked as answer]
There was surprisingly obvious solution to this which I didn't see until now.

You can declare the guards outside of the `guards` key, making them reusable as just regular functions, then your condition can be declared as:

```ts
cond: (...args) => isAuthenticated(...args) && !isPrivateRoute(...args)
```

**Reply by Andarist** [reply]
Power of the language 😉 

That being said - we have "composable" guards in the upcoming v5. So things like `and(['isAuthenticated', not('isPrivateRoute')])` might become possible in the future

**Reply by smac89** [reply]
Just wanted to add a note that I was able to use the lodash [negate](https://lodash.com/docs/#negate) function to achieve the same thing as `not`.

```
cond: negate(isRouteProtected),
```

This is for anyone not willing to write out the full function

**Reply by inakiarroyo** [reply]
In order to reuse the guards functions and also following the idea of the upcoming xstate v5 API naming convention for composing guards you could declare something like:

```js
const guards = {
    isAuthenticated: (context, event) => {
        return ...
    },
    isPrivateRoute: (context, event) => {
        return ...
    },
};

Object.assign(guards, {
    not_isAuthenticated: (...args) => !guards.isAuthenticated(...args),
    not_isPrivateRoute: (...args) => !guards.isPrivateRoute(...args),
    isAuthenticated_and_not_isPrivateRoute: (...args) => guards.isAuthenticated(...args) && !guards.isPrivateRoute(...args)
});
```

For _typescript_  declare `(..args: [Context, Event, GuardMeta<Context, Event>]) => ...` to avoid 
> _A spread argument must either have a tuple type or be passed to a rest parameter_

This allows to mock the guards functions during the testing process.
